Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,207 @@
+            dbgln(OSError(errno).string());
+
+            // TODO: check for root? or 1000
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+Messages::FileSystemAccessServer::PromptSaveFileResponse ClientConnection::prompt_save_file(String const& name, String const& ext, String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+
+    Optional<String> user_picked_file = GUI::FilePicker::get_save_filepath(main_window, name, ext, path_to_view);
+
+    if (user_picked_file.has_value()) {
+        int access = 0;
+
+        if (requested_access & FileAccess::Read)
+            access |= O_RDONLY;
+
+        if (requested_access & FileAccess::Write)
+            access |= O_WRONLY;
+
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        auto fd = open(user_picked_file.value().characters(), access, 0644);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+
+            // TODO: check for root? or 1000
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+void ClientConnection::request_download_file(String const& sender, URL const& url, String const& path)
+{
+    if (!path.starts_with("/"sv)) {
+        // TODO: Make developers give us absolute path
+        // or, take the base path as an argument
+        return;
+    }
+
+    auto main_window = GUI::Window::construct();
+
+    auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} access to download from \"{}\" to \"{}\"?", sender, url.to_string(), path), "Download Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+    if (result == GUI::MessageBox::ExecYes) {
+        auto window = GUI::Window::construct(main_window.ptr());
+        window->resize(300, 150);
+        window->set_title(String::formatted("0% of {}", path));
+        window->set_resizable(false);
+        window->set_main_widget<DownloadWidget>(url, path);
+        window->show();
+    }
+
+    if (result == GUI::MessageBox::ExecNo) {
+        // TODO: If a malicious program requests access over and over, do we eventually want to deny it?
+    }
+}
+
+void ClientConnection::prompt_download_file(URL const& url, String const& filename, String const& ext)
+{
+    auto main_window = GUI::Window::construct();
+
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, filename, ext, Core::StandardPaths::downloads_directory());
+
+    if (user_picked_file.has_value()) {
+        auto window = GUI::Window::construct(main_window.ptr());
```

Comment:
There's more code duplication here, maybe you could have a `open_download_window(URL const& url, String const& path)` method for constructing the window and starting the download.

It could also be a nice idea to add a `DownloadWindow : public GUI::Window` class which has a static method `static void open(URL const& url, String const& path)` or something like that, which does all that stuff for you.